src/methods.jl

```diff
@@ -10,318 +19,451 @@ derivatives.
-    h = x.history
-    if all(p->isdefined(h, p), propertynames(h))
-        @printf io "  roundoff error:        %.2e\n" h.eps
-        @printf io "  bounds on derivatives: %.2e\n" h.bound
-        @printf io "  step size:             %.2e\n" h.step
-        @printf io "  accuracy:              %.2e\n" h.accuracy
-    end
-end
+# Examples
 
-for D in (:Forward, :Backward, :Central, :Nonstandard)
-    @eval begin
-        struct $D{G<:AbstractVector, C<:AbstractVector} <: FiniteDifferenceMethod
-            p::Int
-            q::Int
-            grid::G
-            coefs::C
-            history::History
-        end
-        (d::$D)(f, x=0.0; kwargs...) = fdm(d, f, x; kwargs...)
-    end
-end
+```julia-repl
+julia> fdm = central_fdm(5, 1)
+FiniteDifferenceMethod:
+  order of method:       5
+  order of derivative:   1
+  grid:                  [-2, -1, 0, 1, 2]
+  coefficients:          [0.08333333333333333, -0.6666666666666666, 0.0, 0.6666666666666666, -0.08333333333333333]
 
-# The below does not apply to Nonstandard, as it has its own constructor
-for D in (:Forward, :Backward, :Central)
-    lcname = lowercase(String(D))
-    gridf = Symbol(lcname, "_grid")
-    fdmf = Symbol(lcname, "_fdm")
-
-    @eval begin
-        # Compatibility layer over the "old" API
-        function $fdmf(p::Integer, q::Integer; adapt=1, kwargs...)
-            _dep_kwarg(kwargs)
-            return $D(p, q; adapt=adapt, kwargs...)
-        end
+julia> fdm(sin, 1)
+0.5403023058681155
 
-        function $D(p::Integer, q::Integer; adapt=1, kwargs...)
-            _check_p_q(p, q)
-            grid = $gridf(p)
-            coefs = _coefs(grid, p, q)
-            hist = History(; adapt=adapt, kwargs...)
-            return $D{typeof(grid), typeof(coefs)}(Int(p), Int(q), grid, coefs, hist)
-        end
+julia> fdm(sin, 1) - cos(1)  # Check the error.
+-2.4313884239290928e-14
 
-        @doc """
-            FiniteDifferences.$($(Meta.quot(D)))(p, q; kwargs...)
-            $($(Meta.quot(fdmf)))(p, q; kwargs...)
-
-        Construct a $($lcname) finite difference method of order `p` to compute the `q`th
-        derivative.
-        See [`FiniteDifferenceMethod`](@ref) for more details.
-        """
-        ($D, $fdmf)
-    end
+julia> FiniteDifferences.estimate_step(fdm, sin, 1.0)  # Computes step size and estimates the error.
+(0.0010632902144695163, 1.9577610541734626e-13)
+```
+"""
+function (m::FiniteDifferenceMethod)(
+    f::Function,
+    x::T;
+    factor::Int=1,
+    max_step=convert(T, 0.1)
+) where T<:AbstractFloat
+    # The automatic step size calculation fails if `m.q == 0`, so handle that edge case.
+    iszero(m.q) && return f(x)
+    h, _ = estimate_step(m, f, x, factor=factor, max_step=max_step)
+    return m(f, x, h)
+end
+# Handle arguments that are not floats. Assume that converting to float is desired.
+function (m::FiniteDifferenceMethod)(f::Function, x::T; kw_args...) where T<:Real
+    return m(f, float(x); kw_args...)
```

Comment:
This promotion scheme won't work for e.g. `Unitful` types, where `float(x)` does not give an `AbstractFloat`.  You'll get an infinite dispatch loop leading to a `StackOverflow` exception.

Maybe instead, do something like
```
(m::FiniteDifferenceMethod)(f::Function, x; kw_args...) = _evalrule(m, f, float(x); kw_args...)
```
and then define `_evalrule` to do the actual work.

Note that having `_evalrule` accept any type of `x` will allow many things beyond Unitful arguents. For example, it could also be used to take derivatives for complex arguments — the same finite-difference rules should work perfectly well, assuming the function is analytic at `x`.

Probably also set `max_step = 0.1*oneunit(x)` in `_evalrule`, for a similar reason — `convert` will fail for a unitful type.

Reply:
Possibly we could just change this convience method to only be applies to subtypes of `<:Base.BitInteger` or `Union{Int32, Int64}` or `Int`
It's really just a convience for folk who are being lazy, and are writing something with interger literals.
And we relax the main defintion to accept any `Real`. (then it would work fine on e.g. `Rational`s i think?)
and if people give somehting bad then they can get an error then.

Reply:
What's wrong with my suggestion, i.e. calling `float` on all arguments and then passing it to a duck-typed helper routine?

Reply:
Nothing.
It's also an option.

We could also call float and error if that didn't change the type,
Like SpecialFunctions.jl does.
https://github.com/JuliaMath/SpecialFunctions.jl/blob/master/src/gamma.jl#L551

I think your solution is probably better though than that.